Userland/Libraries/LibPDF/Function.cpp

```diff
@@ -18,11 +18,121 @@ struct Bound {
+    auto bits_per_sample = TRY(document->resolve_to<int>(dict->get_value(CommonNames::BitsPerSample)));
+    switch (bits_per_sample) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+    case 12:
+    case 16:
+    case 24:
+    case 32:
+        // Ok!
+        break;
+    default:
+        dbgln("invalid /BitsPerSample {}", bits_per_sample);
+        return Error { Error::Type::MalformedPDF, "Function type 0 has invalid /BitsPerSample" };
+    }
+
+    Order order = Order::Linear;
+    if (dict->contains(CommonNames::Order))
+        order = static_cast<Order>(TRY(document->resolve_to<int>(dict->get_value(CommonNames::Order))));
+    if (order != Order::Linear && order != Order::Cubic)
+        return Error { Error::Type::MalformedPDF, "Function type 0 has invalid /Order" };
+
+    Vector<Bound> encode;
+    if (dict->contains(CommonNames::Encode)) {
+        auto encode_array = TRY(dict->get_array(document, CommonNames::Encode));
+        if (encode_array->size() % 2 != 0)
+            return Error { Error::Type::MalformedPDF, "Function type 0 /Encode size not multiple of 2" };
+        for (size_t i = 0; i < encode_array->size(); i += 2)
+            encode.append({ encode_array->at(i).to_float(), encode_array->at(i + 1).to_float() });
+    } else {
+        for (unsigned const size : size)
+            encode.append({ 0, static_cast<float>(size - 1) });
+    }
+    if (encode.size() != size.size())
+        return Error { Error::Type::MalformedPDF, "Function type 0 /Encode array has invalid size" };
+
+    Vector<Bound> decode;
+    if (dict->contains(CommonNames::Decode)) {
+        auto decode_array = TRY(dict->get_array(document, CommonNames::Decode));
+        if (decode_array->size() % 2 != 0)
+            return Error { Error::Type::MalformedPDF, "Function type 0 /Decode size not multiple of 2" };
+        for (size_t i = 0; i < decode_array->size(); i += 2)
+            decode.append({ decode_array->at(i).to_float(), decode_array->at(i + 1).to_float() });
+    } else {
+        decode = range.value();
+    }
+    if (decode.size() != range.value().size())
+        return Error { Error::Type::MalformedPDF, "Function type 0 /Decode array has invalid size" };
+
+    size_t size_product = 1;
+    for (unsigned const size : size)
+        size_product *= size;
+    size_t bits_per_plane = size_product * bits_per_sample;
+    size_t total_bits = bits_per_plane * decode.size();
+    if (stream->bytes().size() < ceil_div(total_bits, 8ull))
+        return Error { Error::Type::MalformedPDF, "Function type 0 stream too small" };
+
+    auto function = adopt_ref(*new SampledFunction());
+    function->m_domain = move(domain);
+    function->m_range = move(range.value());
+    function->m_size = move(size);
+    function->m_bits_per_sample = bits_per_sample;
+    function->m_order = order;
+    function->m_encode = move(encode);
+    function->m_decode = move(decode);
```

Comment:
Put these as constructor parameters? IDK, maybe you have reasons not to which I'm too tired to see.

Reply:
Same as the other PR: Adding them in the ctor is one line per init and also one parameter per arg, and it doesn't give any additional encapsulation since this is a factory method on the class.